File: Programmers/C++/메뉴 리뉴얼_3.cpp
#include <string>
#include <vector>
#include <algorithm>
#define FOR(i,n) for(int i=0;i<(n);++i)
#define FORS(i,s,n) for(int i=(s);i<(n);++i)

using namespace std;

vector<string> res;
unordered_map<string, int> um[11];

void dfs(int idx, int cnt, string& s, string& cur) {
    if(idx>s.size()) return;
    if(cur.size()==cnt) {
        ++um[cnt][cur];
        return;
    }
    cur+=s[idx];
    dfs(idx+1,cnt,s,cur);
    cur.pop_back();
    dfs(idx+1,cnt,s,cur);
}

vector<string> solution(vector<string> orders, vector<int> course) {
    
    for(string s: orders) {
        sort(s.begin(),s.end());
        FORS(i,2,s.size()+1) {
            string cur = "";
            dfs(0,i,s,cur);
        }
    }
    
    for(int n: course) {
        int mx = 0;
        for(auto x: um[n]) mx=max(mx,x.second);
        if(mx<=1) continue;
        for(auto x: um[n]) if(mx==x.second) res.push_back(x.first);
    }
    
    sort(res.begin(), res.end());
    
    return res;
}

File: Programmers/Swift/메뉴 리뉴얼_3.swift
import Foundation

var res = [String]()
var hash = [String:Int]()

func dfs(_ idx:Int, _ cnt:Int, _ s:[Character], _ cur: inout [Character]) {
    if cnt == cur.count {
        var str = cur.map{String($0)}.joined()
        var val = 0
        if let num = hash[str] {val=num}
        hash.updateValue(val+1, forKey: str)
        return
    }
    if idx == s.count {return}
    cur.append(s[idx])
    dfs(idx+1,cnt,s,&cur)
    cur.removeLast()
    dfs(idx+1,cnt,s,&cur)
}

func solution(_ orders:[String], _ course:[Int]) -> [String] {
    
    for cnt in course {
        hash.removeAll()
        for i in 0..<orders.count {
            var s = Array(orders[i])
            s.sort()
            var cur = [Character]()
            dfs(0,cnt,s,&cur)
        }
        
        var mx = 0
        for (key, value) in hash {
            mx = max(mx, value)
        }
        if mx <= 1 {continue}
        for (key, value) in hash {
            if mx == value {res.append(key)}
        }
    }
    
    res.sort()
    
    return res
}
